File: src/java/foundgine-core/src/main/java/com/foundgine/core/semantic/SemanticContractSnapshot.java
package com.foundgine.core.semantic;

import com.foundgine.core.abstractions.*;
import java.util.*;

/** Immutable runtime representation of a frozen semantic contract. */
public final class SemanticContractSnapshot {
	private final Map<EntityId, SemanticEntity> entities;
	private final List<SemanticTraversal> traversals;
	private final Map<EntityId, Map<String, Integer>> entityAliasWeights;
	private final Map<EntityId, Map<FieldId, Map<String, Integer>>> fieldAliasWeights;
	private final Map<RelationshipId, Map<String, Integer>> relationshipAliasWeights;
	private final String contractFingerprint;

	public SemanticContractSnapshot(SemanticModel model) {
		model.ensureFrozen();
		entities = Collections.unmodifiableMap(new LinkedHashMap<>(
				model.entities().stream().collect(java.util.stream.Collectors.toMap(SemanticEntity::id,
						java.util.function.Function.identity(), (a, b) -> a, LinkedHashMap::new))));
		traversals = List.copyOf(model.traversals());
		entityAliasWeights = buildEntity();
		fieldAliasWeights = buildField();
		relationshipAliasWeights = buildRelationship();
		contractFingerprint = model.contractFingerprint();
	}

	private Map<EntityId, Map<String, Integer>> buildEntity() {
		var m = new LinkedHashMap<EntityId, Map<String, Integer>>();
		for (var e : entities.values())
			m.put(e.id(), aliasMap(e.effectiveAliases()));
		return Collections.unmodifiableMap(m);
	}

	private Map<EntityId, Map<FieldId, Map<String, Integer>>> buildField() {
		var m = new LinkedHashMap<EntityId, Map<FieldId, Map<String, Integer>>>();
		for (var e : entities.values()) {
			var x = new LinkedHashMap<FieldId, Map<String, Integer>>();
			for (var f : e.fields())
				x.put(f.id(), aliasMap(f.effectiveAliases()));
			m.put(e.id(), Collections.unmodifiableMap(x));
		}
		return Collections.unmodifiableMap(m);
	}

	private Map<RelationshipId, Map<String, Integer>> buildRelationship() {
		var m = new LinkedHashMap<RelationshipId, Map<String, Integer>>();
		for (var e : entities.values())
			for (var r : e.relationships())
				m.put(r.id(), aliasMap(r.effectiveAliases()));
		return Collections.unmodifiableMap(m);
	}

	private static Map<String, Integer> aliasMap(List<SemanticAlias> aliases) {
		var m = new TreeMap<String, Integer>(String.CASE_INSENSITIVE_ORDER);
		for (var a : aliases)
			if (a.weight() != null)
				m.merge(a.name(), a.weight(), Math::max);
		return Collections.unmodifiableMap(m);
	}

	public String contractFingerprint() {
		return contractFingerprint;
	}

	public Collection<SemanticEntity> entities() {
		return entities.values();
	}

	public List<SemanticTraversal> traversals() {
		return traversals;
	}

	public SemanticEntity get(EntityId id) {
		var e = entities.get(id);
		if (e == null)
			throw new IllegalStateException("Entity " + id + " has no semantic descriptor.");
		return e;
	}

	public SemanticEntity resolveEntity(String name) {
		return entities.values().stream()
				.filter(e -> e.name().equalsIgnoreCase(name)
						|| e.effectiveAliases().stream().anyMatch(a -> a.name().equalsIgnoreCase(name)))
				.findFirst()
				.orElseThrow(() -> new NoSuchElementException("Semantic entity '" + name + "' is not defined."));
	}

	public OptionalInt aliasWeight(EntityId id, String alias) {
		return weight(entityAliasWeights.get(id), alias);
	}

	public OptionalInt aliasWeight(EntityId e, FieldId f, String alias) {
		var x = fieldAliasWeights.get(e);
		return weight(x == null ? null : x.get(f), alias);
	}

	public OptionalInt aliasWeight(RelationshipId id, String alias) {
		return weight(relationshipAliasWeights.get(id), alias);
	}

	public boolean tryGetAlias(EntityId id, String alias, java.util.function.IntConsumer sink) {
		var w = aliasWeight(id, alias);
		if (w.isEmpty())
			return false;
		sink.accept(w.getAsInt());
		return true;
	}

	public boolean tryGetAlias(EntityId entityId, FieldId fieldId, String alias, java.util.function.IntConsumer sink) {
		var w = aliasWeight(entityId, fieldId, alias);
		if (w.isEmpty())
			return false;
		sink.accept(w.getAsInt());
		return true;
	}

	public boolean tryGetAlias(RelationshipId relationshipId, String alias, java.util.function.IntConsumer sink) {
		var w = aliasWeight(relationshipId, alias);
		if (w.isEmpty())
			return false;
		sink.accept(w.getAsInt());
		return true;
	}

	private static OptionalInt weight(Map<String, Integer> m, String a) {
		if (m == null)
			return OptionalInt.empty();
		var v = m.get(a);
		return v == null ? OptionalInt.empty() : OptionalInt.of(v);
	}
}

File: src/csharp/Foundgine.Core/Semantic/SemanticContractSnapshot.cs
using System.Collections.ObjectModel;
using Foundgine.Core.Abstractions;

namespace Foundgine.Core.Semantic;

/// <summary>
/// Immutable runtime representation of a validated and frozen semantic contract.
/// A snapshot deliberately separates trusted runtime consumption from the
/// <see cref="SemanticModel"/> construction/lifecycle representation.
/// </summary>
public sealed class SemanticContractSnapshot
{
    private readonly IReadOnlyDictionary<EntityId, SemanticEntity> _entities;
    private readonly IReadOnlyDictionary<EntityId, IReadOnlyDictionary<string, int>> _entityAliasWeights;

    private readonly IReadOnlyDictionary<EntityId, IReadOnlyDictionary<FieldId, IReadOnlyDictionary<string, int>>>
        _fieldAliasWeights;

    private readonly IReadOnlyDictionary<RelationshipId, IReadOnlyDictionary<string, int>> _relationshipAliasWeights;
    private readonly IReadOnlyList<SemanticTraversal> _traversals;

    internal SemanticContractSnapshot(SemanticModel model)
    {
        ArgumentNullException.ThrowIfNull(model);
        model.EnsureFrozen();

        _entities = FreezeEntities(model.Entities);
        _entityAliasWeights = BuildEntityAliasIndex(_entities.Values);
        _fieldAliasWeights = BuildFieldAliasIndex(_entities.Values);
        _relationshipAliasWeights = BuildRelationshipAliasIndex(_entities.Values);
        _traversals = FreezeTraversals(model.Traversals);
        ContractFingerprint = model.ContractFingerprint;
    }

    /// <summary>
    /// The canonical fingerprint of the frozen semantic contract represented by this snapshot.
    /// </summary>
    public string ContractFingerprint { get; }

    /// <summary>All semantic entities in the contract.</summary>
    public IReadOnlyCollection<SemanticEntity> Entities => _entities.Values.ToArray();

    /// <summary>Logical caller-facing traversals in the contract.</summary>
    public IReadOnlyList<SemanticTraversal> Traversals => _traversals;

    public bool TryGet(EntityId id, out SemanticEntity entity) =>
        _entities.TryGetValue(id, out entity!);

    public SemanticEntity Get(EntityId id) =>
        _entities.TryGetValue(id, out var entity)
            ? entity
            : throw new KeyNotFoundException($"Entity {id} has no semantic descriptor.");

    public bool TryResolveEntity(string name, out SemanticEntity entity)
    {
        entity = _entities.Values.FirstOrDefault(x =>
            string.Equals(x.Name, name, StringComparison.OrdinalIgnoreCase) ||
            x.EffectiveAliases.Any(a => string.Equals(a.Name, name, StringComparison.OrdinalIgnoreCase)))!;
        return entity is not null;
    }

    public SemanticEntity ResolveEntity(string name) =>
        TryResolveEntity(name, out var entity)
            ? entity
            : throw new KeyNotFoundException($"Semantic entity '{name}' is not defined.");

    public bool TryGetTraversal(EntityId source, string name, out SemanticTraversal traversal)
    {
        traversal = _traversals.FirstOrDefault(x =>
            x.Source == source && string.Equals(x.Name, name, StringComparison.OrdinalIgnoreCase))!;
        return traversal is not null;
    }

    public SemanticTraversal GetTraversal(EntityId source, string name) =>
        TryGetTraversal(source, name, out var traversal)
            ? traversal
            : throw new KeyNotFoundException(
                $"Semantic traversal '{name}' is not defined on entity '{Get(source).Name}'.");


    public bool TryGetAlias(EntityId entityId, string alias, out int weight) =>
        TryGetWeight(_entityAliasWeights, entityId, alias, out weight);

    public bool TryGetAlias(EntityId entityId, FieldId fieldId, string alias, out int weight)
    {
        if (_fieldAliasWeights.TryGetValue(entityId, out var fields))
            return TryGetWeight(fields, fieldId, alias, out weight);

        weight = default;
        return false;
    }

    public bool TryGetAlias(RelationshipId relationshipId, string alias, out int weight) =>
        TryGetWeight(_relationshipAliasWeights, relationshipId, alias, out weight);

    private static bool TryGetWeight<TKey>(
        IReadOnlyDictionary<TKey, IReadOnlyDictionary<string, int>> index,
        TKey key,
        string alias,
        out int weight)
        where TKey : notnull
    {
        if (index.TryGetValue(key, out var aliases) && aliases.TryGetValue(alias, out weight))
            return true;

        weight = default;
        return false;
    }

    private static IReadOnlyDictionary<EntityId, IReadOnlyDictionary<string, int>> BuildEntityAliasIndex(
        IEnumerable<SemanticEntity> entities) =>
        entities.ToDictionary(
            x => x.Id,
            x => BuildAliasMap(x.EffectiveAliases),
            EqualityComparer<EntityId>.Default);

    private static IReadOnlyDictionary<EntityId, IReadOnlyDictionary<FieldId, IReadOnlyDictionary<string, int>>>
        BuildFieldAliasIndex(
            IEnumerable<SemanticEntity> entities) =>
        entities.ToDictionary(
            e => e.Id,
            e => (IReadOnlyDictionary<FieldId, IReadOnlyDictionary<string, int>>)e.Fields
                .ToDictionary(f => f.Id, f => BuildAliasMap(f.EffectiveAliases)));

    private static IReadOnlyDictionary<RelationshipId, IReadOnlyDictionary<string, int>> BuildRelationshipAliasIndex(
        IEnumerable<SemanticEntity> entities) =>
        entities.SelectMany(x => x.Relationships)
            .ToDictionary(x => x.Id, x => BuildAliasMap(x.EffectiveAliases));

    private static IReadOnlyDictionary<string, int> BuildAliasMap(IEnumerable<SemanticAlias> aliases)
    {
        var map = new Dictionary<string, int>(StringComparer.OrdinalIgnoreCase);
        foreach (var alias in aliases)
        {
            if (alias.Weight is not int weight)
                continue;

            if (!map.TryGetValue(alias.Name, out var existing) || weight > existing)
                map[alias.Name] = weight;
        }

        return new ReadOnlyDictionary<string, int>(map);
    }

    private static IReadOnlyDictionary<EntityId, SemanticEntity> FreezeEntities(
        IEnumerable<SemanticEntity> entities)
    {
        var copy = new Dictionary<EntityId, SemanticEntity>();
        foreach (var entity in entities)
        {
            copy[entity.Id] = entity with
            {
                Fields = new ReadOnlyCollection<SemanticField>(
                    entity.Fields.Select(FreezeField).ToArray()),
                Relationships = new ReadOnlyCollection<SemanticRelationship>(
                    entity.Relationships.Select(FreezeRelationship).ToArray()),
                Aliases = new ReadOnlyCollection<SemanticAlias>(
                    entity.EffectiveAliases.Select(x => new SemanticAlias(x.Name, x.Weight)).ToArray())
            };
        }

        return new ReadOnlyDictionary<EntityId, SemanticEntity>(copy);
    }

    private static SemanticField FreezeField(SemanticField field) => field with
    {
        Aliases = new ReadOnlyCollection<SemanticAlias>(
            field.EffectiveAliases.Select(x => new SemanticAlias(x.Name, x.Weight)).ToArray()),
        Constraints = new ReadOnlyCollection<SemanticConstraint>(
            field.EffectiveConstraints
                .Select(x => new SemanticConstraint(x.Kind, x.Value, x.Minimum, x.Maximum))
                .ToArray())
    };

    private static SemanticRelationship FreezeRelationship(SemanticRelationship relationship) => relationship with
    {
        Aliases = new ReadOnlyCollection<SemanticAlias>(
            relationship.EffectiveAliases.Select(x => new SemanticAlias(x.Name, x.Weight)).ToArray())
    };

    private static IReadOnlyList<SemanticTraversal> FreezeTraversals(
        IEnumerable<SemanticTraversal> traversals)
    {
        var copy = traversals
            .Select(x => new SemanticTraversal(
                x.Source,
                x.Name,
                x.Target,
                new ReadOnlyCollection<RelationshipId>(x.Path.ToArray())))
            .ToArray();

        return new ReadOnlyCollection<SemanticTraversal>(copy);
    }
}